state/protocol/badger/validity.go

```diff
@@ -264,3 +270,96 @@ func IsValidRootSnapshot(snap protocol.Snapshot, verifyResultID bool) error {
 
 	return nil
 }
+
+// IsValidRootSnapshotQCs checks internal consistency of QCs that are included in the root state snapshot
+// It verifies QCs for main consensus and for each collection cluster.
+func IsValidRootSnapshotQCs(snap protocol.Snapshot) error {
+	// validate main consensus QC
+	err := validateRootQC(snap)
+	if err != nil {
+		return fmt.Errorf("invalid root QC: %w", err)
+	}
+
+	// validate each collection cluster separately
+	curEpoch := snap.Epochs().Current()
+	clusters, err := curEpoch.Clustering()
+	if err != nil {
+		return fmt.Errorf("could not get clustering for root snapshot: %w", err)
+	}
+	for clusterIndex := range clusters {
+		cluster, err := curEpoch.Cluster(uint(clusterIndex))
+		if err != nil {
+			return fmt.Errorf("could not get cluster %d for root snapshot: %w", clusterIndex, err)
+		}
+		err = validateClusterQC(cluster)
+		if err != nil {
+			return fmt.Errorf("invalid cluster qc %d: %W", clusterIndex, err)
+		}
+	}
+	return nil
+}
+
+// validateRootQC performs validation of root QC
+// Returns nil on success
+func validateRootQC(snap protocol.Snapshot) error {
+	rootBlock, err := snap.Head()
+	if err != nil {
+		return fmt.Errorf("could not get root block: %w", err)
+	}
+
+	identities, err := snap.Identities(filter.HasRole(flow.RoleConsensus))
+	if err != nil {
+		return fmt.Errorf("could not get root snapshot identities: %w", err)
+	}
+
+	rootQC, err := snap.QuorumCertificate()
+	if err != nil {
+		return fmt.Errorf("could not get root QC: %w", err)
+	}
+
+	dkg, err := snap.Epochs().Current().DKG()
+	if err != nil {
+		return fmt.Errorf("could not get DKG for root snapshot: %w", err)
+	}
+
+	hotstuffRootBlock := model.GenesisBlockFromFlow(rootBlock)
+	committee, err := committees.NewStaticCommitteeWithDKG(identities, flow.Identifier{}, dkg)
+	if err != nil {
+		return fmt.Errorf("could not create static committee: %w", err)
+	}
+	verifier := verification.NewCombinedVerifier(committee, signature.NewConsensusSigDataPacker(committee))
+	forks := &mocks.ForksReader{}
+	hotstuffValidator := validator.New(committee, forks, verifier)
+	err = hotstuffValidator.ValidateQC(rootQC, hotstuffRootBlock)
+	if err != nil {
+		return fmt.Errorf("could not validate root qc: %w", err)
+	}
+	return nil
+}
+
+// validateClusterQC performs QC validation of single collection cluster
+// Returns nil on success
+func validateClusterQC(cluster protocol.Cluster) error {
+	clusterBlock := cluster.RootBlock()
+	clusterRootBlock := &model.Block{
+		BlockID:     clusterBlock.ID(),
+		View:        clusterBlock.Header.View,
+		ProposerID:  clusterBlock.Header.ProposerID,
+		QC:          nil,
+		PayloadHash: clusterBlock.Header.PayloadHash,
+		Timestamp:   clusterBlock.Header.Timestamp,
+	}
```

Comment:
could we [extract it into a function](https://github.com/onflow/flow-go/blob/master/cmd/bootstrap/run/cluster_qc.go#L23-L30), and reuse here?